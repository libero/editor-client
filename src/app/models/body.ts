import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
