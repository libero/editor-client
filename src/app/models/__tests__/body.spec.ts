import { createBodyState } from '../body';

describe('Manuscript state factory', () => {
  it('creates body state', () => {
    const el = document.createElement('body');
    el.innerHTML = 'A sample <italic>text</italic> with <sub>a</sub> variety of <sup>tags</sup>';
    const editorState = createBodyState(el, 'SOME_ID');
    expect(editorState).toMatchSnapshot();
  });
});
