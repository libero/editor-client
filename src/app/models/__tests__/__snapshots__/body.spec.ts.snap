// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manuscript state factory creates body state 1`] = `
Object {
  "doc": Object {
    "content": Array [
      Object {
        "content": Array [
          Object {
            "text": "A sample ",
            "type": "text",
          },
          Object {
            "marks": Array [
              Object {
                "type": "italic",
              },
            ],
            "text": "text",
            "type": "text",
          },
          Object {
            "text": " with ",
            "type": "text",
          },
          Object {
            "marks": Array [
              Object {
                "type": "subscript",
              },
            ],
            "text": "a",
            "type": "text",
          },
          Object {
            "text": " variety of ",
            "type": "text",
          },
          Object {
            "marks": Array [
              Object {
                "type": "superscript",
              },
            ],
            "text": "tags",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
  "selection": Object {
    "anchor": 1,
    "head": 1,
    "type": "text",
  },
}
`;
